fms/utils/generation.py

```diff
@@ -51,13 +121,24 @@ def generate(
         num_beams: TODO: support beam search
         use_cache: requires that the model accept use_cache and
             past_key_value_states args in forward method.
+        pad_id: int
+            the pad token id to use in case of batch generation where sequences are of different lengths
     """
     batched = False
+    requires_padding = False
     if num_beams != 1:
         raise NotImplementedError("generate() does yet not support beam search")
-    if type(input_ids) == torch.Tensor:
+    if isinstance(input_ids, torch.Tensor):
+        # TODO: nested tensors
         if input_ids.dim() != 1:
             batched = True
+    elif isinstance(input_ids, list):
+        max_length = max(p.size(0) for p in input_ids)
+        requires_padding = any(p.size(0) < max_length for p in input_ids)
+        if requires_padding:
```

Comment:
do we actually need "requires_padding"? if the length == max_length then it will just pad with 0 tokens, which will be a no-op, right?

Reply:
ya would basically be a no-op if `requires_padding` is really False and we still do left_pad.

Reply:
I don't see `requires_padding` being used below on attn mask.

Maybe it should be, to preserve the specialized is_causal=true behavior when a special mask isn't needed? but then I might want to call it something like `requires_specialized_mask` or somethiung like that

Reply:
requires_padding is being used to set the pad_id=-1 if no padding was being used, then the internal create_attention_mask decides on the attention mask based on the pad_id. Figured there was no need for the private create_attention_mask to take both the pad_id and the requires_padding. Open to reformatting this though if it makes it more clear.